File: nodecore-miners-pop/src/main/java/nodecore/miners/pop/tasks/ProcessManager.java
// VeriBlock PoP Miner
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.miners.pop.tasks;

import com.google.common.eventbus.Subscribe;
import nodecore.miners.pop.InternalEventBus;
import nodecore.miners.pop.Threading;
import nodecore.miners.pop.contracts.BaseTask;
import nodecore.miners.pop.contracts.PoPMiningOperationState;
import nodecore.miners.pop.contracts.TaskResult;
import nodecore.miners.pop.events.FilteredBlockAvailableEvent;
import nodecore.miners.pop.events.PoPMiningOperationCompletedEvent;
import nodecore.miners.pop.events.TransactionConfirmedEvent;
import nodecore.miners.pop.services.BitcoinService;
import nodecore.miners.pop.services.NodeCoreService;

public class ProcessManager {
    private final NodeCoreService nodeCoreService;
    private final BitcoinService bitcoinService;

    public ProcessManager(NodeCoreService nodeCoreService, BitcoinService bitcoinService) {
        this.nodeCoreService = nodeCoreService;
        this.bitcoinService = bitcoinService;

        InternalEventBus.getInstance().register(this);
    }

    public void shutdown() {
        InternalEventBus.getInstance().unregister(this);
    }

    public void submit(PoPMiningOperationState state) {
        BaseTask task = new GetPoPInstructionsTask(nodeCoreService, bitcoinService);
        Threading.TASK_POOL.submit(() -> executeTask(task, state));
    }

    public void restore(PoPMiningOperationState state) {
        BaseTask task = new RestoreTask(nodeCoreService, bitcoinService);
        Threading.TASK_POOL.submit(() -> executeTask(task, state));
    }

    private void executeTask(BaseTask task, PoPMiningOperationState state) {
        TaskResult result = task.execute(state);
        if (result.isSuccess()) {
            doNext(result.getNext(), result.getState());
        } else {
            handleFail(result.getState());
        }
    }

    private void handleFail(PoPMiningOperationState state) {
        InternalEventBus.getInstance().post(new PoPMiningOperationCompletedEvent(state.getOperationId()));
    }

    private void doNext(BaseTask next, PoPMiningOperationState state) {
        if (next != null) {
            executeTask(next, state);
        }
    }

    @Subscribe
    public void onTransactionConfirmed(TransactionConfirmedEvent event) {
        BaseTask task = new DetermineBlockOfProofTask(nodeCoreService, bitcoinService);
        Threading.TASK_POOL.submit(() -> executeTask(task, event.getState()));
    }

    @Subscribe
    public void onFilteredBlockAvailable(FilteredBlockAvailableEvent event) {
        BaseTask task = new ProveTransactionTask(nodeCoreService, bitcoinService);
        Threading.TASK_POOL.submit(() -> executeTask(task, event.getState()));
    }
}

File: nodecore-miners-pop/src/main/java/nodecore/miners/pop/tasks/ProcessManager.kt
// VeriBlock PoP Miner
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.miners.pop.tasks

import com.google.common.util.concurrent.ThreadFactoryBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.launch
import nodecore.miners.pop.EventBus
import nodecore.miners.pop.core.MiningOperation
import nodecore.miners.pop.services.BitcoinService
import nodecore.miners.pop.services.NodeCoreService
import org.bitcoinj.utils.ContextPropagatingThreadFactory
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class ProcessManager(
    private val nodeCoreService: NodeCoreService,
    private val bitcoinService: BitcoinService
) {
    private val taskPool = Executors.newFixedThreadPool(
        50,
        ThreadFactoryBuilder().apply {
            setNameFormat("pop-tasks-%d")
            setThreadFactory(ContextPropagatingThreadFactory("pop-tasks"))
        }.build()
    )
    private val coroutineScope = CoroutineScope(taskPool.asCoroutineDispatcher())

    init {
        EventBus.transactionSufferedReorgEvent.register(this, ::onReorg)
    }

    fun shutdown() {
        EventBus.transactionSufferedReorgEvent.unregister(this)

        // Shut down threading
        taskPool.shutdown()
        try {
            if (!taskPool.awaitTermination(10, TimeUnit.SECONDS)) {
                taskPool.shutdownNow()
            }
        } catch (ex: InterruptedException) {
            taskPool.shutdownNow()
            Thread.currentThread().interrupt()
        }
    }

    fun submit(operation: MiningOperation) {
        if (operation.job != null) {
            error("Trying to submit operation [${operation.id}] while it already had a running job!")
        }
        operation.job = coroutineScope.launch {
            runTasks(nodeCoreService, bitcoinService, operation)
        }
    }

    private fun onReorg(operation: MiningOperation) {
        // Cancel that mining operation's job
        operation.stopJob()
        // Resubmit it
        submit(operation)
    }
}
